### PR TITLE
Some tweaks to liquid handling

### PR DESCRIPTION
**Pull request by codemime**
Determination of infinite sources of liquid is broken. Some item charges are `int`s and some others are `long`s. It causes a couple of observable bugs:

**Minor (note the number of charges):**
screenshot_2016-05-28_01-06-05

**Major:**
1. Go to an infinite source of water (rivers, lakes etc)
2. Examine it and select "Pour on the ground"
3. Behold a labour of Sisyphus in an infinite loop (finite actually).

The game won't warn you that you can't drink an ocean. It won't ask you where do you prefer to pour it either. Also, pouring liquid into a container is in the menu now:

screenshot_2016-05-28_01-04-12

### CODE REVIEW

File: src/consumption.cpp [hunk 1]
```diff
@@ -381,7 +381,7 @@ edible_rating player::can_eat( const item &food, bool interactive, bool force )
         overfull = !maybe_query( _( "You're full.  Force yourself to eat?" ) );
     } else if( ( ( nutr > 0 && temp_hunger < capacity ) ||
                  ( comest->quench > 0 && temp_thirst < capacity ) ) &&
-               !eathealth && !slimespawner ) {
+               !food.has_infinite_charges() && !eathealth && !slimespawner ) {
```
Comment (BevapDin) [inline src/consumption.cpp]: Why is this relevant here at all? The character is going to eat only one charge anyway.

File: src/item.h [hunk 2]
```diff
@@ -1364,6 +1364,11 @@ class item : public JsonSerializer, public JsonDeserializer, public visitable<it
         * Returns label from "item_label" itemvar and quantity
         */
         std::string label( unsigned int quantity = 0 ) const;
+        /**
+         * Returns true if the item has infinite charges
```
Comment (BevapDin) [inline src/item.h]: The comment only repeats part of the function name and does provide no further information. In fact, it does not even tell what it returns (if at all) when the item does not have infinite charges. Functions like this probably don't need a specific comment at all.

File: src/game.cpp [hunk 3]
```diff
@@ -10584,14 +10592,15 @@ bool game::handle_liquid( item &liquid, item * const source, const int radius,
...
     const size_t chosen = static_cast<size_t>( menu.ret );
     if( chosen >= actions.size() ) {
         add_msg( _( "Never mind." ) );
         // Explicitly canceled all options (container, drink, pour).
         return false;
     }
-    actions[chosen]();
-    return true;
+
+    return actions[chosen]();
```
Comment (BevapDin) [inline src/game.cpp]: This kind of changes the meaning of the return value, please make sure it still fits the documented eaning (see game.h).

Returning true means the player wanted to handle the liquid, but may have failed to actually do anything with it (e.g. choose to pour it into a container, but could not find a suitable one, or choose to pour into a keg on the ground, but that was already full). It should return `false` when the player explicitly canceled.

**Comment by BevapDin** [comment]
> Also, pouring liquid into a container is in the menu now:

I pondered that change as well. But I assumed player will usually (mostly?) want to fill a container item, which would require one keystroke (the inventory letter) only. Now it requires two keystrokes. It's more consistent and allows players to see the alternative actions. Again, I'm torn on that.

File: src/consumption.cpp (hunk 1 above)
Comment (codemime) [inline src/consumption.cpp]: It prevents the redundant query below. It's obvious that the character won't be able to drink up an infinite source of water. If they have chosen the "Consume it" in the menu, they mean it.

File: src/game.cpp (hunk 3 above)
Comment (codemime) [inline src/game.cpp]: Thanks. I'll double check this.

**Comment by codemime** [comment]
When I was playing the game for the first time and tried to pick up some stuff from a toilet (faugh!) I was quite surprised. The game asked me to chose a container for the toilet water, then it insisted that I'll have to drink it from my hands. Only when I had rejected all that offers, it showed me the items that I wanted. I think one menu with all available options is the most natural way of interacting and two keystrokes is a necessary evil. Reasonable defaults can help mitigate it.

**Comment by Podesta** [comment]
Maybe show the containers you can pour it in the prompt? This way it remains consistent without increasing keystrokes for the most used action.

**Comment by codemime** [comment]
@Podesta This inevitably would lead to conflicts. Imagine you have a container that's assigned to `e` or `g`.

**Comment by Coolthulhu** [comment]
The best way would be to somehow integrate extra choices into the inventory menu, though that would require designing some consistent, workable way to make other features use that too.

A relatively usable but ugly solution would be to have the available containers listed in the menu (the one with "drink" and "pour" choices) in addition to the "pour into container" option. This would allow one keypress pouring, except in the cases where the container is bound to a reserved character (one used for drinking, pouring on the ground, filling vehicle or pouring into container).
But this would be ugly since there would be 2 menus sharing functionality.

**Comment by Coolthulhu** [comment]
Doesn't gray out the option to pour into container when no container is available.

Doesn't gray out the option to pour infinite source onto the ground.

**Comment by codemime** [comment]
> Doesn't gray out the option to pour into container when no container is available.

This one is tricky for now. The functionality is encapsulated in the `inventory_selector` class. I plan further refactoring there.

> Doesn't gray out the option to pour infinite source onto the ground.

I had done it, but then decided that an explicit message is more informative. Maybe the option should be grayed out nevertheless.

**Pull request merged**